File: algorithms/2_medium/M/03946_maximum_number_of_items_from_sale_i/dp.c
static inline int max_int(const int a, const int b)
{
    return a > b ? a : b;
}

static int max_factor(int** items, const int items_len)
{
    int res = INT_MIN;
    for (int i = 0; i < items_len; i += 1) {
        assert(items[i][0] >= 1 && items[i][0] <= 1500);
        res = max_int(res, items[i][0]);
    }
    return res;
}


int maximumSaleItems(int** items, const int items_len, const int* items_cols_len, const int budget)
{
    assert(budget >= 1 && budget<= 1500 && items_len > 0 &&
           items != NULL && items_cols_len != NULL && items_cols_len[0] == 2);
    const int max = max_factor(items, items_len);
    int* buff = calloc((size_t)(2 * (max + 1) + 2 * (budget + 1)), sizeof(*buff));
    int* extra = &buff[0];
    int* freqs = &buff[1 + max];
    int* curr = &buff[2 * (1 + max)];
    int* next = &buff[2 * (1 + max) + 1 + budget];

    for (int i = 0; i < items_len; i += 1) {
        freqs[items[i][0]] += 1;
    }

    for (int i = 1; i <= max; i += 1) {
        int curr = i;
        while (curr <= max) {
            extra[i] += freqs[curr];
            curr += i; 
        }
    }

    for (int i = 0; i < items_len; i += 1) {
        const int price_i = items[i][1];
        if (price_i > budget) {
            continue;
        }
        const int cnt = extra[items[i][0]];
        for (int j = price_i; j <= budget; j += 1) {
            next[j] = max_int(next[j], curr[j - price_i] + cnt);
            next[j] = max_int(next[j], next[j - price_i] + 1);
        }
        for (int idx = 1; idx <= budget; idx += 1) {
            curr[idx] = next[idx];
        }
    }

/* cleanup: */
    const int res = curr[budget];
    free(buff);
    return res; 
}

File: algorithms/2_medium/M/03946_maximum_number_of_items_from_sale_i/dp.rs
impl Solution {
    pub fn maximum_sale_items(items: Vec<Vec<i32>>, budget: i32) -> i32 {
        assert!(budget >= 1 && budget<= 1500);
        assert!(items.iter().all(|x| x[0] >= 1 && x[0] <= 1500));
        let max = items.iter().map(|x| x[0]).max().unwrap();
        let items_len = items.len();
        let mut buff = vec![0_i32; (2 * (max + 1) + 2 * (budget + 1)) as usize];
        let (mut extra, mut remain_buff) = buff.split_at_mut(1 + max as usize);
        let (mut freqs, mut remain_buff) = remain_buff.split_at_mut(1 + max as usize);
        let (mut curr, mut next) = remain_buff.split_at_mut((budget + 1) as usize);

        for i in 0..items_len {
            freqs[items[i][0] as usize] += 1;
        }

        for i in 1..=max {
            let mut curr = i;
            while curr <= max {
                extra[i as usize] += freqs[curr as usize];
                curr += i; 
            }
        }

        for i in 0..items_len {
            let price_i = items[i][1];
            if price_i > budget {
                continue;
            }
            let cnt = extra[items[i][0] as usize];
            for j in price_i..=budget {
                next[j as usize] = curr[j as usize].max(curr[(j - price_i) as usize] + cnt)
                                                   .max(next[(j - price_i) as usize] + 1);
            }
            for idx in 0..curr.len() {
                curr[idx] = next[idx];
            }
        }

        curr[budget as usize]
    }
}
